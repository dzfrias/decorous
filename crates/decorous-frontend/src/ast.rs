use rslint_parser::SyntaxNode;

#[derive(Debug)]
pub struct DecorousAst<'a> {
    nodes: Vec<Node<'a, Location>>,
    script: Option<SyntaxNode>,
    css: Option<&'a str>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Node<'a, T> {
    metadata: T,
    node_type: NodeType<'a, T>,
}

#[derive(Debug, PartialEq, Eq, Clone, Hash, Copy)]
pub struct Location {
    start: usize,
    end: usize,
}

#[derive(Debug, Clone, PartialEq)]
pub enum NodeType<'a, T> {
    Element(Element<'a, T>),
    Text(&'a str),
    Comment(&'a str),
    SpecialBlock(SpecialBlock<'a, T>),
    Mustache(SyntaxNode),
    Error,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Element<'a, T> {
    tag: &'a str,
    attrs: Vec<Attribute<'a>>,
    children: Vec<Node<'a, T>>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum SpecialBlock<'a, T> {
    For(ForBlock<'a, T>),
    If(IfBlock<'a, T>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct ForBlock<'a, T> {
    binding: &'a str,
    index: Option<&'a str>,
    expr: SyntaxNode,
    inner: Vec<Node<'a, T>>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct IfBlock<'a, T> {
    expr: SyntaxNode,
    inner: Vec<Node<'a, T>>,
    // TODO: Implement
    else_block: Option<Vec<Node<'a, T>>>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Attribute<'a> {
    EventHandler(EventHandler<'a>),
    Binding(&'a str),
    KeyValue(&'a str, Option<AttributeValue<'a>>),
}

#[derive(Debug, Clone, PartialEq)]
pub struct EventHandler<'a> {
    event: &'a str,
    expr: SyntaxNode,
}

#[derive(Debug, Clone, PartialEq)]
pub enum AttributeValue<'a> {
    Literal(&'a str),
    JavaScript(SyntaxNode),
}

impl<'a, T> Element<'a, T> {
    pub fn new(tag: &'a str, attrs: Vec<Attribute<'a>>, children: Vec<Node<'a, T>>) -> Self {
        Self {
            tag,
            attrs,
            children,
        }
    }

    pub fn tag(&self) -> &str {
        self.tag
    }

    pub fn children(&self) -> &[Node<'_, T>] {
        self.children.as_ref()
    }

    pub fn attrs(&self) -> &[Attribute<'_>] {
        self.attrs.as_ref()
    }

    pub fn children_mut(&mut self) -> &mut Vec<Node<'a, T>> {
        &mut self.children
    }
}

impl<'a, T> Node<'a, T> {
    pub fn new(ty: NodeType<'a, T>, metadata: T) -> Self {
        Self {
            metadata,
            node_type: ty,
        }
    }

    pub fn node_type(&self) -> &NodeType<'a, T> {
        &self.node_type
    }

    pub fn node_type_mut(&mut self) -> &mut NodeType<'a, T> {
        &mut self.node_type
    }

    pub fn metadata(&self) -> &T {
        &self.metadata
    }

    pub fn metadata_mut(&mut self) -> &mut T {
        &mut self.metadata
    }

    pub fn error(metadata: T) -> Self {
        Self::new(NodeType::Error, metadata)
    }

    pub fn cast_meta<U, F>(self, transfer_func: &mut F) -> Node<'a, U>
    where
        F: FnMut(T) -> U,
    {
        match self.node_type {
            NodeType::SpecialBlock(block) => Node {
                metadata: transfer_func(self.metadata),
                node_type: NodeType::SpecialBlock(match block {
                    SpecialBlock::If(if_block) => SpecialBlock::If(IfBlock {
                        expr: if_block.expr,
                        inner: if_block
                            .inner
                            .into_iter()
                            .map(|node| node.cast_meta(transfer_func))
                            .collect(),
                        else_block: if_block.else_block.map(|nodes| {
                            nodes
                                .into_iter()
                                .map(|node| node.cast_meta(transfer_func))
                                .collect()
                        }),
                    }),
                    SpecialBlock::For(for_block) => SpecialBlock::For(ForBlock {
                        inner: for_block
                            .inner
                            .into_iter()
                            .map(|node| node.cast_meta(transfer_func))
                            .collect(),
                        binding: for_block.binding,
                        index: for_block.index,
                        expr: for_block.expr,
                    }),
                }),
            },
            NodeType::Element(elem) => Node {
                metadata: transfer_func(self.metadata),
                node_type: NodeType::Element(Element {
                    tag: elem.tag,
                    attrs: elem.attrs,
                    children: elem
                        .children
                        .into_iter()
                        .map(|child| child.cast_meta(transfer_func))
                        .collect(),
                }),
            },
            NodeType::Text(text) => Node {
                metadata: transfer_func(self.metadata),
                node_type: NodeType::Text(text),
            },
            NodeType::Comment(comment) => Node {
                metadata: transfer_func(self.metadata),
                node_type: NodeType::Comment(comment),
            },
            NodeType::Mustache(syntax_node) => Node {
                metadata: transfer_func(self.metadata),
                node_type: NodeType::Mustache(syntax_node),
            },
            NodeType::Error => Node {
                metadata: transfer_func(self.metadata),
                node_type: NodeType::Error,
            },
        }
    }
}

impl Location {
    pub fn new(start: usize, end: usize) -> Self {
        Self { start, end }
    }

    pub fn char(idx: usize) -> Self {
        Self {
            start: idx,
            end: idx,
        }
    }

    pub fn start(&self) -> usize {
        self.start
    }

    pub fn end(&self) -> usize {
        self.end
    }
}

impl<'a, T> ForBlock<'a, T> {
    pub fn new(
        binding: &'a str,
        index: Option<&'a str>,
        expr: SyntaxNode,
        inner: Vec<Node<'a, T>>,
    ) -> Self {
        Self {
            binding,
            index,
            expr,
            inner,
        }
    }

    pub fn binding(&self) -> &str {
        self.binding
    }

    pub fn index(&self) -> Option<&str> {
        self.index
    }

    pub fn expr(&self) -> &SyntaxNode {
        &self.expr
    }

    pub fn inner(&self) -> &[Node<'_, T>] {
        self.inner.as_ref()
    }

    pub fn inner_mut(&mut self) -> &mut Vec<Node<'a, T>> {
        &mut self.inner
    }
}

impl<'a> EventHandler<'a> {
    pub fn new(event: &'a str, expr: SyntaxNode) -> Self {
        Self { event, expr }
    }

    pub fn event(&self) -> &str {
        self.event
    }

    pub fn expr(&self) -> &SyntaxNode {
        &self.expr
    }
}

impl<'a, T> IfBlock<'a, T> {
    pub fn new(
        expr: SyntaxNode,
        inner: Vec<Node<'a, T>>,
        else_block: Option<Vec<Node<'a, T>>>,
    ) -> Self {
        Self {
            expr,
            inner,
            else_block,
        }
    }

    pub fn expr(&self) -> &SyntaxNode {
        &self.expr
    }

    pub fn inner(&self) -> &[Node<'a, T>] {
        self.inner.as_ref()
    }

    pub fn else_block(&self) -> Option<&[Node<'a, T>]> {
        self.else_block.as_ref().map(|block| block.as_slice())
    }

    pub fn inner_mut(&mut self) -> &mut Vec<Node<'a, T>> {
        &mut self.inner
    }
}

impl<'a> DecorousAst<'a> {
    pub fn new(
        nodes: Vec<Node<'a, Location>>,
        script: Option<SyntaxNode>,
        css: Option<&'a str>,
    ) -> Self {
        Self { nodes, script, css }
    }

    pub fn nodes(&self) -> &[Node<'_, Location>] {
        self.nodes.as_ref()
    }

    pub fn script(&self) -> Option<&SyntaxNode> {
        self.script.as_ref()
    }

    pub fn css(&self) -> Option<&'a str> {
        self.css
    }

    pub fn iter_nodes(&'a self) -> NodeIter<'a, Location> {
        let nodes = self.nodes().iter().collect::<Vec<&'a Node<'a, Location>>>();
        NodeIter::new(nodes)
    }

    pub fn into_components(self) -> (Vec<Node<'a, Location>>, Option<SyntaxNode>, Option<&'a str>) {
        (self.nodes, self.script, self.css)
    }
}

#[derive(Debug)]
pub struct NodeIter<'a, T> {
    stack: Vec<&'a Node<'a, T>>,
}

impl<'a, T> NodeIter<'a, T> {
    fn new(node: Vec<&'a Node<'a, T>>) -> NodeIter<'a, T> {
        let mut stack = Vec::new();
        stack.extend(node);
        Self { stack }
    }
}

impl<'a, T> Iterator for NodeIter<'a, T> {
    type Item = &'a Node<'a, T>;

    fn next(&mut self) -> Option<Self::Item> {
        self.stack.pop().map(|node| {
            if let NodeType::Element(elem) = node.node_type() {
                self.stack.extend(elem.children().iter().rev());
            }

            node
        })
    }
}
