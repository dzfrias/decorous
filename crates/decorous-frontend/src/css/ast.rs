use std::{borrow::Cow, fmt};

use itertools::Itertools;
use rslint_parser::SyntaxNode;

#[derive(Debug)]
pub struct Css<'a> {
    rules: Vec<Rule<'a>>,
}

#[derive(Debug, PartialEq)]
pub enum Rule<'a> {
    At(AtRule<'a>),
    Regular(RegularRule<'a>),
}

#[derive(Debug, PartialEq)]
pub struct RegularRule<'a> {
    selector: Selector<'a>,
    declarations: Vec<Declaration<'a>>,
}

#[derive(Debug, PartialEq)]
pub struct AtRule<'a> {
    name: &'a str,
    additional: &'a str,
    contents: Option<Vec<Rule<'a>>>,
}

#[derive(Debug, PartialEq)]
pub struct Selector<'a> {
    parts: Vec<SelectorPart<'a>>,
}

#[derive(Debug, PartialEq)]
pub struct SelectorPart<'a> {
    text: Option<Cow<'a, str>>,
    pseudoes: Vec<Pseudo<'a>>,
}

#[derive(Debug, PartialEq)]
pub enum Pseudo<'a> {
    Element(&'a str),
    Class {
        name: &'a str,
        value: Option<&'a str>,
    },
}

#[derive(Debug, PartialEq)]
pub struct Declaration<'a> {
    name: &'a str,
    values: Vec<Value<'a>>,
}

#[derive(Debug, PartialEq)]
pub enum Value<'a> {
    Mustache(SyntaxNode),
    Css(&'a str),
}

impl<'a> Declaration<'a> {
    pub fn new(name: &'a str, values: Vec<Value<'a>>) -> Self {
        Self { name, values }
    }

    pub fn name(&self) -> &str {
        self.name
    }

    pub fn values(&self) -> &[Value<'_>] {
        self.values.as_ref()
    }

    pub fn values_mut(&mut self) -> &mut Vec<Value<'a>> {
        &mut self.values
    }

    pub fn name_mut(&mut self) -> &mut &'a str {
        &mut self.name
    }
}

impl<'a> SelectorPart<'a> {
    pub fn new(text: Option<Cow<'a, str>>, pseudoes: Vec<Pseudo<'a>>) -> Self {
        Self { text, pseudoes }
    }

    pub fn text(&self) -> Option<&Cow<'a, str>> {
        self.text.as_ref()
    }

    pub fn pseudoes(&self) -> &[Pseudo<'_>] {
        self.pseudoes.as_ref()
    }

    pub fn text_mut(&mut self) -> &mut Option<Cow<'a, str>> {
        &mut self.text
    }

    pub fn pseudoes_mut(&mut self) -> &mut Vec<Pseudo<'a>> {
        &mut self.pseudoes
    }
}

impl<'a> Selector<'a> {
    pub fn new(parts: Vec<SelectorPart<'a>>) -> Self {
        Self { parts }
    }

    pub fn parts_mut(&mut self) -> &mut Vec<SelectorPart<'a>> {
        &mut self.parts
    }

    pub fn parts(&self) -> &[SelectorPart<'_>] {
        self.parts.as_ref()
    }
}

impl<'a> RegularRule<'a> {
    pub fn new(selector: Selector<'a>, declarations: Vec<Declaration<'a>>) -> Self {
        Self {
            selector,
            declarations,
        }
    }

    pub fn selector(&self) -> &Selector<'a> {
        &self.selector
    }

    pub fn declarations(&self) -> &[Declaration<'_>] {
        self.declarations.as_ref()
    }

    pub fn selector_mut(&mut self) -> &mut Selector<'a> {
        &mut self.selector
    }

    pub fn declarations_mut(&mut self) -> &mut Vec<Declaration<'a>> {
        &mut self.declarations
    }
}

impl<'a> Css<'a> {
    pub fn new(rules: Vec<Rule<'a>>) -> Self {
        Self { rules }
    }

    pub fn rules(&self) -> &[Rule<'a>] {
        &self.rules
    }

    pub fn rules_mut(&mut self) -> &mut Vec<Rule<'a>> {
        &mut self.rules
    }
}

impl<'a> AtRule<'a> {
    pub fn new(name: &'a str, additional: &'a str, contents: Option<Vec<Rule<'a>>>) -> Self {
        Self {
            name,
            additional,
            contents,
        }
    }

    pub fn name(&self) -> &str {
        self.name
    }

    pub fn additional(&self) -> &str {
        self.additional
    }

    pub fn contents(&self) -> Option<&[Rule<'a>]> {
        self.contents.as_deref()
    }

    pub fn contents_mut(&mut self) -> &mut Option<Vec<Rule<'a>>> {
        &mut self.contents
    }

    pub fn additional_mut(&mut self) -> &mut &'a str {
        &mut self.additional
    }

    pub fn name_mut(&mut self) -> &mut &'a str {
        &mut self.name
    }
}

// ---Display impls---

impl<'a> fmt::Display for SelectorPart<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}{}",
            self.text().unwrap_or(&Cow::default()),
            self.pseudoes().iter().join("")
        )
    }
}

impl<'a> fmt::Display for Value<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Mustache(s) => write!(f, "{s}"),
            Self::Css(css) => write!(f, "{css}"),
        }
    }
}

impl<'a> fmt::Display for Pseudo<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Element(elem) => write!(f, "::{elem}"),
            Self::Class {
                name,
                value: Some(value),
            } => write!(f, ":{name}({value})"),
            Self::Class { name, value: None } => {
                write!(f, ":{name}")
            }
        }
    }
}

impl<'a> fmt::Display for Declaration<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}: {};", self.name, self.values().iter().join(" "))
    }
}

impl<'a> fmt::Display for AtRule<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        if let Some(contents) = self.contents() {
            write!(
                f,
                "@{} {} {{ {} }}",
                self.name(),
                self.additional(),
                contents.iter().join(" ")
            )
        } else {
            write!(f, "@{} {};", self.name(), self.additional())
        }
    }
}

impl<'a> fmt::Display for RegularRule<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{} {{ {} }}",
            self.selector(),
            self.declarations().iter().join("; ")
        )
    }
}

impl<'a> fmt::Display for Rule<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::At(at_rule) => write!(f, "{at_rule}"),
            Self::Regular(regular) => write!(f, "{regular}"),
        }
    }
}

impl<'a> fmt::Display for Selector<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.parts().iter().join(" "))
    }
}

impl<'a> fmt::Display for Css<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.rules().iter().join("\n"))
    }
}
