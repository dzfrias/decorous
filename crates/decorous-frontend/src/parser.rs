use std::{
    collections::{HashMap, VecDeque},
    str::Chars,
};
use thiserror::Error;

use crate::ast::{Element, Location, Node, NodeType};

#[derive(Debug)]
pub struct Parser<'a> {
    // Source
    source: &'a str,
    input: Chars<'a>,
    // Not using `Peekable` because access to `as_str()` of `input` is crucial.
    peek_buffer: Option<char>,

    // A queue we maintain of the closed tags we find. This allows us to find invalid closing tag
    // errors
    close_tag_queue: VecDeque<&'a str>,

    // Current character index
    index: usize,
    // Current character index, with proper utf8 length taken into account. We need this because
    // we take a slice of `source`, which operates in bytes, not code points
    slice_index: usize,

    errors: Vec<ParseError>,
}

#[derive(Debug, Error, PartialEq)]
pub enum ParseError {
    #[error("invalid closing tag, expected {0}")]
    InvalidClosingTag(String, Location),
    #[error("invalid character, expected {0}, got {1}")]
    ExpectedCharacter(char, char, Location),
}

/// Zero-copy parser that takes in decorous HTML syntax.
impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        Self {
            source: input,
            input: input.chars(),
            peek_buffer: None,

            close_tag_queue: VecDeque::new(),

            index: 0,
            slice_index: 0,

            errors: vec![],
        }
    }

    /// Parse the received input into a collection of nodes. Also, return the errors found while
    /// parsing.
    pub fn parse(mut self) -> (Vec<Node<'a>>, Vec<ParseError>) {
        let nodes = self.parse_nodes();

        (nodes, self.errors)
    }

    fn parse_nodes(&mut self) -> Vec<Node<'a>> {
        let mut nodes = Vec::new();

        while self.peek().is_some() {
            // We want to parse '\n', so keep it.
            self.consume_while(|c| c.is_whitespace() && c != '\n');

            match self.peek() {
                Some('<') => {
                    // Should be the '<' we just peeked
                    self.consume();

                    let peek = self.peek();
                    if peek.is_some_and(|c| c == '/') {
                        // Should be the '/' we just peeked.
                        self.consume();
                        self.skip_whitespace();

                        let close = self.parse_tag();

                        self.skip_whitespace();
                        let c = self.consume();
                        if c != Some('>') {
                            self.errors.push(ParseError::ExpectedCharacter(
                                '>',
                                c.unwrap_or('\0'),
                                Location::new(self.index, self.index),
                            ));
                        }

                        // Add to the closing tag queue
                        self.close_tag_queue.push_back(close);
                        break;
                    } else {
                        let node = self.parse_element();

                        // If no closing tag was found (<br/>, for example), just push the tag
                        if self.close_tag_queue.is_empty() {
                            nodes.push(node);
                            continue;
                        }

                        let NodeType::Element(elem) = node.node_type() else {
                            unreachable!("should never parse anything other than an element")
                        };

                        // Validate that the closing tag matches the opening one
                        let tag = elem.tag();
                        let Some(close_tag) = self.close_tag_queue.pop_front() else {
                            unreachable!("checked if empty at top of block");
                        };
                        if close_tag != tag {
                            self.errors
                                .push(ParseError::InvalidClosingTag(tag.to_owned(), node.loc()));
                            continue;
                        }

                        nodes.push(node);
                    }
                }
                _ => {
                    // Parse text node if not tag
                    nodes.push(self.parse_text_node());
                }
            }
        }

        nodes
    }

    fn consume_while<F>(&mut self, predicate: F) -> &'a str
    where
        F: Fn(char) -> bool,
    {
        let start = self.slice_index;
        while self.peek().is_some_and(|c| predicate(c)) {
            self.consume();
        }

        // Get the text we just consumed. Importantly, we use `self.slice_index`, as we're indexing
        // an &str, which means we could invalidate byte boundaries if we just counted by
        // `self.index`
        &self.source[start..self.source.len() - self.source[self.slice_index..].len()]
    }

    fn consume(&mut self) -> Option<char> {
        if let Some(peek) = self.peek_buffer {
            self.index += 1;
            self.slice_index += peek.len_utf8();
            self.peek_buffer = None;
            return Some(peek);
        }

        let c = self.input.next();
        if let Some(c) = c {
            self.index += 1;
            // Slice index should keep track of the total bytes we've consumed.
            self.slice_index += c.len_utf8();
        }
        c
    }

    fn peek(&mut self) -> Option<char> {
        if let Some(peek) = self.peek_buffer {
            return Some(peek);
        }

        let next = self.input.next()?;

        self.peek_buffer = Some(next);
        Some(next)
    }

    fn skip_whitespace(&mut self) -> &'a str {
        self.consume_while(|c| c.is_whitespace())
    }

    fn parse_text_node(&mut self) -> Node<'a> {
        self.consume_while(|c| c.is_whitespace() && c != '\n');
        let mut start = self.index;
        let mut text = self.consume_while(|c| c != '<');
        if text.get(0..=0).is_some_and(|c| c == "\n") {
            start = self.index.saturating_sub(1);
            text = &text[..1];
        }

        let node_type = NodeType::Text(text);
        Node::new(
            node_type,
            Location::new(start, self.index.saturating_sub(1)),
        )
    }

    fn parse_tag(&mut self) -> &'a str {
        // `is_digit(36)` checks if it's a-z, A-Z, or 0-9
        self.consume_while(|c| c.is_digit(36))
    }

    fn parse_element(&mut self) -> Node<'a> {
        let start = self.index;
        let tag = self.parse_tag();
        self.skip_whitespace();
        let attrs = self.parse_attrs();
        let children = self.parse_nodes();

        Node::new(
            NodeType::Element(Element::new(tag, attrs, children)),
            Location::new(start - 1, self.index.saturating_sub(1)),
        )
    }

    fn parse_attrs(&mut self) -> HashMap<&'a str, Option<&'a str>> {
        let mut attrs = HashMap::new();

        while self.peek().is_some_and(|c| c != '>') {
            self.skip_whitespace();
            let name = self.consume_while(|c| !is_control_or_delim(c));
            self.skip_whitespace();
            let value = if self.peek().is_some_and(|c| c == '=') {
                self.consume();
                self.skip_whitespace();
                let s = self.parse_attr_value();
                self.skip_whitespace();
                Some(s)
            } else {
                None
            };

            attrs.insert(name, value);
        }

        // Should be the '>'
        self.consume();
        attrs
    }

    fn parse_attr_value(&mut self) -> &'a str {
        match self.peek() {
            Some(c) if c == '"' || c == '\'' => {
                self.consume();
                let v = self.consume_while(|str_char| str_char != c);
                self.consume();
                v
            }
            _ => {
                todo!("parse mustache tag");
            }
        }
    }
}

fn is_control_or_delim(ch: char) -> bool {
    match ch {
        '\u{007F}' => true,
        c if c >= '\u{0000}' && c <= '\u{001F}' => true,
        c if c >= '\u{0080}' && c <= '\u{009F}' => true,
        ' ' | '"' | '\'' | '>' | '/' | '=' => true,
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! insta_test_all {
        ($($input:expr),+) => {
            $(
                let parser = Parser::new($input);
                let (tree, errs) = parser.parse();
                assert_eq!(Vec::<ParseError>::new(), errs);
                ::insta::assert_debug_snapshot!(tree);
             )+
        };
    }

    macro_rules! insta_test_all_err {
        ($($input_and_errs:expr),+) => {
            $(
                let parser = Parser::new($input_and_errs.0);
                let (tree, errs) = parser.parse();
                assert_eq!($input_and_errs.1, errs);
                ::insta::assert_debug_snapshot!(tree);
             )+
        };
    }

    #[test]
    fn can_parse_text_node() {
        insta_test_all!("hello", "   hello", "hello hello", "你", "你好");
    }

    #[test]
    fn can_parse_basic_elements() {
        insta_test_all!(
            "<p>你好</p>",
            "<tagname><p>inner</p></tagname>",
            "<p>
                <p>hello</p>
            </p>
            "
        );
    }

    #[test]
    fn parses_basic_attrs() {
        insta_test_all!(
            "<p attr1='hello'></p>",
            "<p attr=\"attribute\"></p>",
            "<p ></p>",
            "<p attr=\"你好\"></p>",
            "<p attr=\"one\"     attr2=\"two\">text</p>"
        );
    }

    #[test]
    fn reports_unmatched_tag_errors() {
        insta_test_all_err!((
            "<p></invalid>",
            vec![ParseError::InvalidClosingTag(
                "p".to_owned(),
                Location::new(0, 12)
            )]
        ));
    }

    #[test]
    fn continues_parsing_if_unmatched_tag_found() {
        insta_test_all_err!(
            (
                "<p></invalid>
            <span>hello</span>",
                vec![ParseError::InvalidClosingTag(
                    "p".to_owned(),
                    Location::new(0, 12)
                )],
            ),
            (
                "
                <span>hello
                <p></p>
                <span></span>
                ",
                Vec::<ParseError>::new()
            )
        );
    }

    #[test]
    fn reports_invalid_closing_tag_angle_bracket() {
        insta_test_all_err!((
            "<p></p!",
            vec![ParseError::ExpectedCharacter('>', '!', Location::new(7, 7))]
        ));
    }
}
