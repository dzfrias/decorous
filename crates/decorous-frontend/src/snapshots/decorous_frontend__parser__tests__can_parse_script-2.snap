---
source: crates/decorous-frontend/src/parser.rs
expression: tree
---
DecorousAst {
    nodes: [
        Node {
            metadata: Location {
                start: 23,
                end: 34,
            },
            node_type: Element(
                Element {
                    tag: "p",
                    attrs: [],
                    children: [
                        Node {
                            metadata: Location {
                                start: 26,
                                end: 30,
                            },
                            node_type: Text(
                                "hello",
                            ),
                        },
                    ],
                },
            ),
        },
    ],
    script: Some(
        SCRIPT@0..6
          EXPR_STMT@0..6
            ASSIGN_EXPR@0..5
              NAME_REF@0..1
                IDENT@0..1 "x"
              WHITESPACE@1..2 " "
              EQ@2..3 "="
              WHITESPACE@3..4 " "
              LITERAL@4..5
                NUMBER@4..5 "3"
            SEMICOLON@5..6 ";"
        ,
    ),
    css: None,
}
